src/box/replication.cc

```diff
@@ -896,6 +909,19 @@ applier_on_connect_f(struct trigger *trigger, void *event)
 		break;
 	case APPLIER_CONNECTED:
 		state->connected++;
+		if (bootstrap_strategy == BOOTSTRAP_STRATEGY_SUPERVISED &&
+		    (replicaset_state == REPLICASET_BOOTSTRAP ||
+		     replicaset_state == REPLICASET_JOIN)) {
+			/*
+			 * We may reuse the on_connect trigger, because it fires
+			 * only once. The applier will remain paused after the
+			 * trigger run.
+			 */
+			trigger_clear(trigger);
+			trigger_create(trigger, applier_on_ballot_update_f,
+				       applier, NULL);
+			trigger_add(&applier->on_ballot_update, trigger);
+		}
```

Comment:
Nice, nice. Thanks for doing that! The next step would be to drop some counters from `struct replicaset` like `applier.connected`, `applier.loading`, etc. I think they are used only when some applier changes its state, which means the needed counters could be just re-calculated from scratch in the relevant triggers. But this can be done some time later.